File: services/identity/src/main/java/com/datn/identity/infrastructure/web/FileStorageClient.java
package com.datn.identity.infrastructure.web;

import com.fasterxml.jackson.databind.JsonNode;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.*;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;

import java.util.Map;

@Component
public class FileStorageClient {
    private final RestTemplate restTemplate;
    private final String fileStorageBaseUrl;

    public FileStorageClient(
            RestTemplate restTemplate,
            @Value("${file-storage.base-url:http://file-storage-api:3000}") String fileStorageBaseUrl) {
        this.restTemplate = restTemplate;
        this.fileStorageBaseUrl = fileStorageBaseUrl;
    }

    public record CreatePresignedUrlRequest(
            String originalName,
            String mimeType,
            Long size,
            String service,
            String modelType,
            String subjectId,
            String uploadedBy,
            java.util.List<String> tags,
            Map<String, Object> metadata
    ) {}

    public record CreatePresignedUrlResponse(
            String assetId,
            String presignedUrl,
            String objectKey,
            Integer expiresIn
    ) {}

    public CreatePresignedUrlResponse createPresignedUrl(CreatePresignedUrlRequest request) {
        String url = fileStorageBaseUrl + "/files/presigned-url";

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);

        HttpEntity<CreatePresignedUrlRequest> entity = new HttpEntity<>(request, headers);

        try {
            ResponseEntity<JsonNode> response = restTemplate.exchange(
                    url,
                    HttpMethod.POST,
                    entity,
                    JsonNode.class
            );

            JsonNode body = response.getBody();
            if (body == null) {
                throw new RuntimeException("File-storage service returned empty response");
            }

            JsonNode data = body.get("data");
            if (data == null) {
                throw new RuntimeException("File-storage service response missing 'data' field");
            }

            return new CreatePresignedUrlResponse(
                    data.get("assetId").asText(),
                    data.get("presignedUrl").asText(),
                    data.get("objectKey").asText(),
                    data.get("expiresIn").asInt()
            );
        } catch (org.springframework.web.client.ResourceAccessException e) {
            throw new RuntimeException("Cannot connect to file-storage service at " + fileStorageBaseUrl + ". Please check if the service is running and the URL is correct.", e);
        } catch (Exception e) {
            throw new RuntimeException("Failed to create presigned URL: " + e.getMessage(), e);
        }
    }

    public void confirmUpload(String assetId) {
        String url = fileStorageBaseUrl + "/files/confirm-upload";

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);

        Map<String, String> body = Map.of("assetId", assetId);
        HttpEntity<Map<String, String>> entity = new HttpEntity<>(body, headers);

        restTemplate.exchange(
                url,
                HttpMethod.POST,
                entity,
                Void.class
        );
    }

    public record PresignedGetUrlRequest(
            String id,
            Integer expirySeconds
    ) {}

    public record PresignedGetUrlResponse(
            String id,
            String presignedUrl,
            Integer expiresIn
    ) {}

    public PresignedGetUrlResponse getPresignedGetUrl(String fileId, Integer expirySeconds) {
        String url = fileStorageBaseUrl + "/files/presigned-get-url";

        HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_JSON);

        PresignedGetUrlRequest request = new PresignedGetUrlRequest(fileId, expirySeconds);
        HttpEntity<PresignedGetUrlRequest> entity = new HttpEntity<>(request, headers);

        try {
            ResponseEntity<JsonNode> response = restTemplate.exchange(
                    url,
                    HttpMethod.POST,
                    entity,
                    JsonNode.class
            );

            JsonNode body = response.getBody();
            if (body == null) {
                throw new RuntimeException("File-storage service returned empty response");
            }

            JsonNode data = body.get("data");
            if (data == null) {
                throw new RuntimeException("File-storage service response missing 'data' field");
            }

            return new PresignedGetUrlResponse(
                    data.get("id").asText(),
                    data.get("presignedUrl").asText(),
                    data.get("expiresIn").asInt()
            );
        } catch (org.springframework.web.client.ResourceAccessException e) {
            throw new RuntimeException("Cannot connect to file-storage service at " + fileStorageBaseUrl + ". Please check if the service is running and the URL is correct.", e);
        } catch (Exception e) {
            throw new RuntimeException("Failed to get presigned GET URL: " + e.getMessage(), e);
        }
    }
}

File: services/pm/src/common/file-storage/file-storage.client.ts
import { Injectable, HttpException, HttpStatus } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';

export interface CreatePresignedUrlRequest {
  originalName: string;
  mimeType: string;
  size: number;
  service: string;
  modelType: string;
  subjectId: string;
  uploadedBy?: string;
  orgId?: string;
  tags?: string[];
  metadata?: Record<string, any>;
}

export interface CreatePresignedUrlResponse {
  assetId: string;
  presignedUrl: string;
  objectKey: string;
  expiresIn: number;
}

export interface FileMetadata {
  id: string;
  originalName: string;
  mimeType: string;
  size: number;
  bucket: string;
  objectKey: string;
  url?: string;
  service: string;
  modelType: string;
  subjectId: string;
  uploadedBy?: string;
  tags?: string[];
  metadata?: Record<string, any>;
  uploadStatus: 'pending' | 'completed' | 'failed';
  createdAt: Date;
  updatedAt: Date;
}

export interface PresignedGetUrlResponse {
  id: string;
  presignedUrl: string;
  expiresIn: number;
}

@Injectable()
export class FileStorageClient {
  private readonly baseUrl: string;

  constructor(private readonly configService: ConfigService) {
    this.baseUrl = this.configService.get<string>('FILE_STORAGE_URL', 'http://localhost:41111');
  }

  async createPresignedUrl(dto: CreatePresignedUrlRequest): Promise<CreatePresignedUrlResponse> {
    const response = await fetch(`${this.baseUrl}/files/presigned-url`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(dto),
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new HttpException(
        error.message || 'Failed to create presigned URL',
        response.status || HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }

    const result = await response.json();
    return result.data;
  }

  async confirmUpload(assetId: string): Promise<FileMetadata> {
    const response = await fetch(`${this.baseUrl}/files/confirm-upload`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ assetId }),
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new HttpException(
        error.message || 'Failed to confirm upload',
        response.status || HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }

    const result = await response.json();
    return result.data;
  }

  async getFile(id: string): Promise<FileMetadata> {
    const response = await fetch(`${this.baseUrl}/files/${id}`);

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new HttpException(
        error.message || 'File not found',
        response.status || HttpStatus.NOT_FOUND,
      );
    }

    const result = await response.json();
    return result.data;
  }

  async getPresignedGetUrl(id: string, expirySeconds = 3600): Promise<PresignedGetUrlResponse> {
    const response = await fetch(`${this.baseUrl}/files/presigned-get-url`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ id, expirySeconds }),
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new HttpException(
        error.message || 'Failed to get presigned URL',
        response.status || HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }

    const result = await response.json();
    return result.data;
  }

  async deleteFile(id: string): Promise<void> {
    const response = await fetch(`${this.baseUrl}/files/${id}`, {
      method: 'DELETE',
    });

    if (!response.ok) {
      const error = await response.json().catch(() => ({}));
      throw new HttpException(
        error.message || 'Failed to delete file',
        response.status || HttpStatus.INTERNAL_SERVER_ERROR,
      );
    }
  }
}
